### PR TITLE
LBR unwinding method processing enhancement for Disassembly.

### PR DESCRIPTION
**Pull request by dknysh**
Hello, Milian.

Could you please review the changes to add LBR unwinding method proper processing for Disassembly. 

Earlier we have been starting to discuss the issue:
https://github.com/KDAB/perfparser/pull/14#issuecomment-700005368

There are some required frames are not looked up for LBR unwinding method.
Entered additional disasmFrames to accumulate frames for LBR
to compute then Disassembly events costs within function/method.

Thank you,
Darya

### CODE REVIEW

File: app/perfunwind.cpp [hunk 1]
```diff
@@ -441,10 +441,16 @@ void PerfUnwind::resolveCallchain()
...
 
-    auto reportIp = [&](quint64 ip) -> bool {
+    auto reportIp = [&](quint64 ip, bool branchStack) -> bool {
         symbols->attachDwfl(&m_currentUnwind);
-        m_currentUnwind.frames.append(symbols->lookupFrame(ip, isKernel,
-                                            &m_currentUnwind.isInterworking));
+        int frame = symbols->lookupFrame(ip, isKernel, &m_currentUnwind.isInterworking);
+        // There are some required frames are not looked up for LBR unwinding method.
+        // Entered disasmFrames to accumulate frames for LBR to compute then Disassembly events costs within function/method.
+        if (branchStack) {
```
Comment (milianw) [inline app/perfunwind.cpp]: as I mentioned in the past - I do not understand why we have to distinguish between disasm frames and non-disasm frames. a frame is a frame, either it applies to all views or none.

please expand your commit message with an example that shows why this is needed.

**Review by milianw — CHANGES_REQUESTED** [review]
hey, sorry for the delay. please explain why this is needed, and why we need different frame storage. also please add a testcase somehow, i.e. run perf with lbr on one of the existing static example binaries and try to reproduce it that way

**Comment by dknysh** [comment]
Hello, Milian.
I have updated source code and generated the similar logs for perf.lbr.data recorded with LBR on perfdata/vector_static_gcc/vector_static_gcc_v9.1.0.

[hotspot_generate_script_output.txt](https://github.com/KDAB/perfparser/files/6154562/hotspot_generate_script_output.txt)
[hotspot_generate_script_output_prev.txt](https://github.com/KDAB/perfparser/files/6154563/hotspot_generate_script_output_prev.txt)
[perf_annotate_output.txt](https://github.com/KDAB/perfparser/files/6154564/perf_annotate_output.txt)
[perf_script_output.txt](https://github.com/KDAB/perfparser/files/6154565/perf_script_output.txt)

I also attach diff with current KDAB to show the issue - for example, 0x41b477 __hypot_finite that is missed on version without presented patch. 
diff hotspot_generate_script_output.txt  hotspot_generate_script_output_prev.txt 
[hypot_finite.diff.txt](https://github.com/KDAB/perfparser/files/6154629/hypot_finite.diff.txt)

Could you please review?

**Comment by milianw** [comment]
Thanks for the files, that is very helpful. Thing is, I'm not convinced at all that this is something we need to change.

Please compare these snippets:

`hotspot_generate_script_output_prev.txt`:
```
vector_static_g	3509027	7037283.204239728:	930396 cycles
	000000000041b3b0 __hypot_finite (vector_static_gcc_v9.1.0)
	0000000000417a53 hypot (vector_static_gcc_v9.1.0)
```

`hotspot_generate_script_output.txt`:
```
vector_static_g	3509027	7037283.204239728:	930396 cycles
	000000000041b3b0 __hypot_finite (vector_static_gcc_v9.1.0)
	000000000041b3b4 __hypot_finite (vector_static_gcc_v9.1.0)
	000000000041b3b0 __hypot_finite (vector_static_gcc_v9.1.0)
	0000000000417a53 hypot (vector_static_gcc_v9.1.0)
```

This actually duplicates the address `41b3b0` which is very odd. Is this a jump/loop in the code that is recorded this way?

`perf_script_output.txt`:
```
vector_static_g 3509027 7037283.204239:     930396 cycles: 
	          41b3b0 __ieee754_hypot+0x0 (/home/dknysh/kdab-disasm-fraames/hotspot/3rdparty/perfparser/tests/auto/perfdata/vector_static_gcc/vector_static_gcc_v9.1.0)
	          417a53 __hypot+0x13 (/home/dknysh/kdab-disasm-fraames/hotspot/3rdparty/perfparser/tests/auto/perfdata/vector_static_gcc/vector_static_gcc_v9.1.0)
```

Note how this actually matches the `_prev` output above, so imo that's what we should keep.

`grep 41b3b4 perf_annotate_output.txt`:
```
    0.00 :   41b3b4:       push   %r13
```

So as we can see here, even `perf annotate` doesn't associate any cost with this address that we are supposedly missing.

Once again: I'm not convinced about this patch at all. Please explain why you think we should include it.

Now let's look at the address that you are pointing out:

`hotspot_generate_script_output_prev.txt`:
```
vector_static_g	3509027	7037283.210067432:	720467 cycles
	000000000041b3b0 __hypot_finite (vector_static_gcc_v9.1.0)
	0000000000417a53 hypot (vector_static_gcc_v9.1.0)
```

`hotspot_generate_script_output.txt`:
```
vector_static_g	3509027	7037283.210067432:	720467 cycles
	000000000041b477 __hypot_finite (vector_static_gcc_v9.1.0)
	000000000041b47f __hypot_finite (vector_static_gcc_v9.1.0)
	000000000041b3b0 __hypot_finite (vector_static_gcc_v9.1.0)
```

OK, no duplication in this case - quite the contrary, it actually looks like we are using the wrong address for the final sample location, i.e. the frame points at 0x41b3b0. With this patch it now points at 0x41b477.

`perf_script_output.txt`:
```
vector_static_g 3509027 7037283.210067:     720467 cycles: 
	          41b3b0 __ieee754_hypot+0x0 (/home/dknysh/kdab-disasm-fraames/hotspot/3rdparty/perfparser/tests/auto/perfdata/vector_static_gcc/vector_static_gcc_v9.1.0)
	          417a53 __hypot+0x13 (/home/dknysh/kdab-disasm-fraames/hotspot/3rdparty/perfparser/tests/auto/perfdata/vector_static_gcc/vector_static_gcc_v9.1.0)
```

Hm but it looks like `perf script` also points at 41b3b0 like we are doing...

`grep -E "41b3b0|41b477|41b47f" perf_annotate_output.txt`:
```
   17.26 :   41b3b0:       endbr64
   13.36 :   41b477:       xorpd  0x7f051(%rip),%xmm3        # 49a4d0 <conversion_rate+0xb8>
    0.00 :   41b47f:       movq   %rdx,%xmm4
```

So this is odd indeed, but note how 41b47f still has a cost of 0 - so adding that one to the list of locations is apparently wrong. I believe there's something else going on here that isn't understood properly yet. Please look into this in more detail and explain what perf annotate is doing compared to perf state, then put that knowledge into the commit message.

**Comment by dknysh** [comment]
I am adding 2 additional frames from callchain to branchStack. 
I guess we should expect non-zero costs for the first - the deepest.
But if condition for the first - (!addedUserFrames... does not work, the second will be added.
 
Indeed when we the traverse frames then in hotspot part we look only the first and consider its symbol as visited after seeing.
The only I can suggest - to move that check and 'break' one frame upper and add 1 additional frame instead of 2.

**Comment by milianw** [comment]
> I am adding 2 additional frames from callchain to branchStack.
> I guess we should expect non-zero costs for the first - the deepest.
> But if condition for the first - (!addedUserFrames... does not work, the second will be added.

I'm sorry, but this just says that you (and me too!) don't understand what's going on here. Blindly changing the code just to say "here, looks better" without understanding the background is not something that I find acceptable. If you care about fixing this situation, please investigate what the reason is, then lay out the explanation in the fix.

Are we misunderstanding the LBR chain? If so why and how? Can you link to references in the perf code upstream?

> Indeed when we the traverse frames then in hotspot part we look only the first and consider its symbol as visited after seeing.

Yes, and that is correct for the purpose of hotspot's other views from what I can see. Note how we agree with `perf stat` from what I can see, so that part seems to be correct. But I'm totally unsure if it's only the raw annotation view that is affected. If the LBR chain somehow allows us to distribute the cost of a sample across multiple addresses, e.g. for jumps in loops or similar, then this would also apply to the per-line cost map.

> The only I can suggest - to move that check and 'break' one frame upper and add 1 additional frame instead of 2.

No that is not sufficient. Please try to understand what is going on here and then get back with an explanation.

**Comment by dknysh** [comment]
> Are we misunderstanding the LBR chain? If so why and how? Can you link to references in the perf code upstream?

Previously I have tried to compare perfparser's LBR branchStack bypass and algorithm with perf sources (https://github.com/torvalds/linux/blob/master/tools/perf/util/machine.c   - resolve_lbr_callchain_sample).
But I did not find any suspicious.
perfparser traversal of lbr branchStack does quite the same.
(Except that lbr_callchain_add_stitched_lbr_ip for the moment)

> > The only I can suggest - to move that check and 'break' one frame upper and add 1 additional frame instead of 2.
> 
> No that is not sufficient. Please try to understand what is going on here and then get back with an explanation.

My explanation is the following -
there is the difference for the first element of lbr branchStack - only its entry.to we reportIp.
And it is the deepest callee.
For the other we look only entry.from - caller.
According to the logs and debugger it (that entry.to) points to the start of that deepest callee - __hypot_finite + 0x0.
Probably we should not expect from lbr branchStack more exact address inside that callee but we still need it for Disassembly costs. It is the reason why I wanted to accumulate them into separate frames list.

Then I can suppose that condition
 // sometimes it skips the first user frame.
            if (!addedUserFrames
can be actual for LBR and hasBranchStack  also but I can not proove.

I have committed version with one additional frame and regenerated log and diff with current KDAB version where you can see that all of added addresses have non-zero costs. 
Sorry, I have no better explanation.
[hotspot_generate_script_output.txt](https://github.com/KDAB/perfparser/files/6164061/hotspot_generate_script_output.txt)
[hypot_finite.diff.txt](https://github.com/KDAB/perfparser/files/6164062/hypot_finite.diff.txt)

**Comment by milianw** [comment]
ok cool, this new version looks much simpler and together with the test I'm fine with merging it.

but when you write:
> Probably we should not expect from lbr branchStack more exact address inside that callee but we still need it for Disassembly costs. It is the reason why I wanted to accumulate them into separate frames list.

If you think that based on the LBR it should be possible to distribute the sample cost across multiple adresses (versus just the call chain), then we need a different approach. I.e. that would mean we would basically have to split the single sample into multiple ones and then distribute the cost across them. It should *not* be a separate frames list!

**Comment by dknysh** [comment]
Ok, thanks. Should I upstream the changes to Qt?

**Comment by milianw** [comment]
yes please, always upstream everything

**Comment by dknysh** [comment]
gerrit Change-Id: Ie6413415c573e659505a4715978c65fce135d979

**Pull request merged**